wmfdata/utils.py

```diff
@@ -175,19 +175,30 @@ def sql_tuple(i):
     > sql_tuple(["a", "b", "c"])
     "('a', 'b', 'c')"
     """
+
+    # Transform other iterables into lists, raising errors for non-iterables
+    if type(i) != list:
+        i = [x for x in i]
+
     # It might seem useful to return "()" when an empty iterable is passed, but "IN ()"
     # results in an SQL syntax error. Instead, we should send a clear signal to the caller
     # that it needs to better handle the no-items case.
+    #
+    # Running this check after converting i to a list avoids errors from generators that
+    # don't implement len.
     if len(i) == 0:
         raise ValueError("Cannot produce an SQL tuple without any items.")
 
-    # Transform other iterables into lists, raising errors for non-iterables
-    if type(i) != list:
-        i = [x for x in i]
-
     # Using Python's string representation functionality means we get escaping for free. Using the
     # representation of a tuple almost works, but fails when there's just one element, because SQL
     # doesn't accept the trailing comma that Python uses. Instead, we use representation of a
     # list and replace the brackets with parentheses.
+    #
+    # To-do: The string representation method fails in an edge case. It normally produces single
+    # quotes but falls back to double quotes if the string contains a single quote. If the database
+    # strictly follows the ANSI SQL standard (as Presto does), double-quoted strings are
+    # interpreted as literal column names rather than strings. This is low priority, but ideally
+    # we would fix this by constructing the SQL tuple manually rather than using the string
+    # representation.
     list_repr = repr(i)
     return "(" + list_repr[1:-1] + ")"
```

Comment:
You could avoid the edge case by doing instead:

```
>>> iter = ['enwiki', 'eswiki', 'ptwiki']
>>> quoted_iter = [f"'{e}'" for e in iter]
>>> '(' + ', '.join(quoted_iter) + ')'
"('enwiki', 'eswiki', 'ptwiki')"
```

Reply:
Unfortunately, that doesn't work! One component of the edge case is that the string contains a single quote (`'`). With this, you're wrapping in single quotes, so another single quote in the middle will break it early and cause problems.

Since you nerd-sniped me (😁), I started looking more deeply into what needs to be escaped, and it turns out the problem is even bigger than I realized: there's no single escaping strategy that works for all of MariaDB, Spark, and Presto, so for this to work correctly in all cases, we'll have to implement more than one and let the user choose. More details: https://github.com/nshahquinn/misc-wikimedia-analysis/blob/master/2022-10_SQL_string_escaping.ipynb

Much as I want to, I can't justify spending more time on this now, so I think the best strategy is just to stick with the current implementation. As far as I can tell, it always works as long no single quotes or backlashes get involved, and even then it works about half the time. I will update the docstring and comments with what I've learned, though.